fix(core): not invoking object's toString when rendering to the DOM

crisbeto opened the pull request:
Currently we convert objects to strings using `'' + value` which is quickest, but it stringifies the value using its `valueOf`, rather than `toString`. These changes switch to using `String(value)` which has identical performance and calls the `toString` method as expected. Note that another option was calling `toString` directly, but benchmarking showed it to be slower.

I've included the benchmark I used to verify the performance so we have it for future reference and we can reuse it when making changes to `renderStringify` in the future.

Also for reference, here are the results of the benchmark:

```
Benchmark: renderStringify
 concat: 2.006 ns(0%)
 concat with toString: 2.201 ns(-10%)
 toString: 237.494 ns(-11741%)
 toString with toString: 121.072 ns(-5937%)
 constructor: 2.201 ns(-10%)
 constructor with toString: 2.201 ns(-10%)
 toString mono: 14.536 ns(-625%)
 toString with toString mono: 9.757 ns(-386%)
```

Fixes #38839.

AndrewKushnir commented on packages/core/test/acceptance/text_spec.ts:
```diff
@@ -129,4 +129,26 @@ describe('text instructions', () => {
     expect(div.innerHTML).toBe('function foo() { }');
   });
+
+  it('should stringify an object using its toString method', () => {
```
I believe this change also resolves the problem when binding value is a `Symbol`. This is not very practical, but it used to work in VE and is currently failing with Ivy (and this PR should resolve that). If you get a chance, could you please add a quick test for the `Symbol` as a value? Thank you.

crisbeto replied:
Ah, that's right. I'd forgotten that we talked about the `Symbol` case. I've added another unit test for it.

AndrewKushnir replied:
Awesome, thanks a lot @crisbeto 👍

crisbeto committed:
fix(core): not invoking object's toString when rendering to the DOM

Currently we convert objects to strings using `'' + value` which is quickest,
but it stringifies the value using its `valueOf`, rather than `toString`. These
changes switch to using `String(value)` which has identical performance
and calls the `toString` method as expected. Note that another option
was calling `toString` directly, but benchmarking showed it to be slower.

I've included the benchmark I used to verify the performance so we have it
for future reference and we can reuse it when making changes to `renderStringify`
in the future.

Also for reference, here are the results of the benchmark:

```
Benchmark: renderStringify
 concat: 2.006 ns(0%)
 concat with toString: 2.201 ns(-10%)
 toString: 237.494 ns(-11741%)
 toString with toString: 121.072 ns(-5937%)
 constructor: 2.201 ns(-10%)
 constructor with toString: 2.201 ns(-10%)
 toString mono: 14.536 ns(-625%)
 toString with toString mono: 9.757 ns(-386%)
```

Fixes #38839.

mhevery commented:
[presubmit](http://test/OCL:344847216:BASE:344847229:1606762902899:2e265751)

mhevery commented:
[presubmit](https://tap.corp.google.com/ui#id=OCL:344847216:BASE:344847229:1606762903019:35f4f854)

AndrewKushnir commented:
FYI, adding "risk: medium" to indicate that it might make sense to land it as a separate CL in g3.

mhevery reviewed (approved):
reviewed-for: global-approvers
